### ISSUE TITLE
New Primitive "Dynamic Symbols" / "Function Symbols" 

### ISSUE THREAD

**Issue by fizykz** [typst/typst#2698]
### Description

Hopefully this isn't too long... 😅 I wanted to make sure it was absolutely clear. 

Often when typesetting, you find that you often have to type a lot of variations of the same general concept.  You need a derivative?  What kind, total or partial?  First or second order?  Should it be with respect to `x` or `t`?  Are we differentiating `f` or `zeta`?   Arguably, one of the top goals of any type setter should be to make this as easy of a process as possible.  To that end, I'd like to propose an idea that helps the problem above. 
The idea is to do something similar to how `symbol` works:
```
arrow \\ Right arrow
arrow.l \\ Left arrow
arrow.l.r.double // Double headed bold arrow
```
`symbol` is special because
1) It behaves like a function call without being an explicit call.  E.g. typing `arrow` alone is enough to print one.
2) It behaves like a dictionary/tree, allowing indices to determine the behavior while also making it look like "chaining."  This was a great idea of Typst to add as it allows avoiding countless arguments and prevents there from being several functions.  It just looks like one "bundle," which is `arrow`. 

My proposal is to add a primitive like `symbol`, except instead of strings/content, it can also accept functions (which most likely return strings/content).  This is similar to `__call__` but it doesn't need to be explicitly called to give content.  Below will be many examples ranging from very helpful tidbits, and (possibly bold) ways to styling your document code.  What I think is best about this feature though is the precedent for it.  `symbol` is already in, and for many of us, that's what we're used to.  I think adding this allows the possibility of Typst's style to be more homogenous and uniform.

Also, these should work even outside of math mode. 

### Use Case

To demonstrate this, for each example I'll first show what it might be equivalent to in terms of a function call, and then I'll show various examples, giving an idea of how it looks.  I'll give an example of how such an object could be declared at the end, but I think multiple valid ways are reasonable.  The implementation doesn't really matter so long as the feel is the same in use. 

Admittedly, there's one major flaw with these arguments, which is that, with enough effort, one can easily make a `symbol` instance which handles every one of these cases without dynamically handling it.  However I have two arguments to this:
1) There could be countless cases which is bothersome to add independently by hand to a `symbol`.  Think of all the different variables we could differentiate with respect to.  Like `x_i`. 
2) There can still be ambiguous cases.  Consider `u` and `v` which are often used both to differentiate with respect to, and differentiate as is in differential geometry.  Should `derivative.v` be the derivative of `v`?  Or the derivative with respect to `v`?  It isn't clear, and even if we use a convention, how should we label the alternative choice?  
With this, I show some examples:

__Usage:__

__Derivative operator w.r.t. `x`__
Currently:
```derivative() \\ OR
derivative(variable: x)
```
Equivalent proposals:
```
derivative
derivative.variable(t)
```

__Partial derivative operator w.r.t. `t`__
Currently:
```
derivative(variable: t, partial: true) 
derivative_partial(variable: t)
```
Equivalent proposals:
```
derivative.partial.t
derivative.t.partial
derivative.partial.variable(t)
derivative.variable(t).partial
```

__Partial derivative of `F` w.r.t. `t`__
Currently:
```
derivative(F, t, partial: true) 
derivative_partial(F, t)
```
Equivalent proposals:
```
derivative.partial.t.F
derivative(F, t).partial
derivative.partial.variable(t).function(F)
```

__Second partial derivative of `f` with respect to `x`__
Currently:
```
derivative(F, t, partial: true, order: 2) 
derivative_partial(F, t, order: 2)
```
Equivalent proposals:
```
derivative.partial.t.f.second
derivative(r, t).partial.second
derivative(f, t, 2).partial
derivative(f, t).partial.order(2)
```

With this, here's an example of how we might construct this dynamic symbol in a way that's similar to `symbol`.  This could easily have its own syntax which would be much easier to follow however. 

__Construction__
Currently:
```
#let derivative(function: f, variable: x, partial: false, order: 1) {
\\ ... or something along those lines
} 
```
Proposal:
```
#let function(flags, func, var, order) {
  \\ Construct and return the symbol
  \\ `flags` is a dictionary of boolean values, corresponding to the construction of the dynamic symbol.
} 

#let derivative = dynamic_symbol(
  ("func", "var", "order"), 
  ("partial"), // There's no second element because it's meant to be used as a boolean flag in the function. 
  ("order", "order"), // The argument passed to .order is the "order" argument. 
  ("function", "func"), // Likewise as above
  ("curl"), // Nabla cross
  ("t"), // Replaces "var" argument with "t" 
  ("f"),
  ("prime"), 
  ("parenthetical") \\ short for   f^((order)) 
  ("newton") \\ for Newton dot notation! 
  function
) 
```

We could do a lot more with this, for example I put "curl", dot notation, and more under it, as it feels well grouped with it.  And as said, we could have some custom syntax such that flagging (i.e. writing `derivstive.f`) `f` automatically sets `function` to `"f" `.  This helps with error detection & catches common values (since we rarely differentiate with respect to `f`), helping users, and it's syntax reduces weight on package developers to check a lot of "short cut" flags (like `f`, `t`, and `x`) 

Finally, the last thing I'd like to mention is it might also be nice if this could just return the value that the dynamic symbol uses.  For example we could then uses this like an environment/grouping:

```
// Declare `result` dynamic symbol (Like Lemmas, corollaries, theorems, propositions, examples)
// Instead of returning content, it returns a dictionary with methods. 

#let TypstTheorem = result.theorem[
  Typst is very cool and dynamic! 
].proof[
  Visit Typst.app.  The rest of the proof is left to the reader.
].name("Typst's Theorem") 

Lorum Ipsum x500

= Chapter 13:  Why Typst is Great
We come to a major result:
TypstTheorem.setLocation() // Sets the marked location as being in chalter 13.
TypstTheorem.display()  // Displays something like "Theorem 13.1 (Typst's Theorem):" 
Here's the proof:
TypstTheorem.display_proof()

= Chapter 14
TypstTheorem.display() // Again displays "Theorem 13.1", possibly even linking it to its original placement.
```

**Comment by Leedehai** [comment]
FWIW, (ordinary, partial) derivatives and differentials are available in the [phyisca](https://github.com/Leedehai/typst-physics) package.
<img width="822" alt="Screenshot 2023-11-28 at 12 53 12" src="https://github.com/typst/typst/assets/18319900/9cd92974-bda8-472b-b906-3fe6a6e2b90b">

**Comment by wmstack** [comment]
I agree that we need a new Typst primitive. The regular standard library has objects that act as
1. A symbol
2. A function
3. A dictionary

There is basically no way for the user to have an object which acts like that and is convenient to use. For example, Typst's `dash` symbol, you can use it by itself, and it outputs a dash. It is also a dictionary, and has multiple modifiers that you can apply to it. It also accepts a parameter as input at puts a horizontal bar above it.


Similarly, a `hat` acts as a symbol as well as a function. There has to be a way for a user to create a similar symbol that can be used standalone, or act as a function or a dictionary at the same time.
